File: backend/app/schemas/api_automation.py
from datetime import datetime
from typing import Any, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field


class ApiRequestDetails(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    path: str
    method: str = "POST"
    url_params: dict[str, Any] = Field(default_factory=dict)
    form: dict[str, Any] = Field(default_factory=dict)
    body_json: dict[str, Any] = Field(
        default_factory=dict,
        validation_alias=AliasChoices("json", "body_json"),
        serialization_alias="json",
    )
    cookies: dict[str, Any] = Field(default_factory=dict)
    headers: dict[str, Any] = Field(default_factory=dict)


class ApiAutomationCase(BaseModel):
    id: int
    project_id: Optional[int] = None
    module_id: Optional[int] = None
    module_name: str
    exec_type: str = "HTTP 请求"
    priority: int = Field(ge=1, le=5)
    name: str
    request: ApiRequestDetails
    expected: str
    created_at: Optional[datetime] = None


class ApiAutomationCaseQuery(BaseModel):
    project_id: Optional[int] = None
    name: Optional[str] = None
    priority: Optional[int] = Field(default=None, ge=1, le=5)
    module_id: Optional[int] = None
    exec_type: Optional[str] = None


class ApiExecutionResult(BaseModel):
    case_id: int
    case_name: str
    status: str
    expected: str
    ai_record: str
    response: dict[str, Any]
    request: ApiRequestDetails


class ApiExecutionDetails(BaseModel):
    summary: dict[str, int]
    results: list[ApiExecutionResult]


class ApiTestExecBase(BaseModel):
    project_id: Optional[int] = None
    name: str = Field(min_length=1, max_length=255)
    exec_type: str = Field(default="HTTP 请求", max_length=255)
    case_ids: list[int] = Field(min_length=1)
    details: Optional[ApiExecutionDetails] = None
    desc: Optional[str] = Field(default=None, max_length=1024)
    exec_param: dict[str, Any] = Field(default_factory=dict)
    exec_status: str = Field(default="已完成", max_length=255)


class ApiTestExecCreate(ApiTestExecBase):
    pass


class ApiTestExecUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    exec_status: Optional[str] = Field(default=None, max_length=255)
    details: Optional[ApiExecutionDetails] = None
    desc: Optional[str] = Field(default=None, max_length=1024)
    exec_param: Optional[dict[str, Any]] = None


class ApiTestExecOut(ApiTestExecBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ApiTestExecRunRequest(BaseModel):
    project_id: Optional[int] = None
    name: str = Field(min_length=1, max_length=255)
    exec_type: str = Field(default="HTTP 请求", max_length=255)
    case_ids: list[int] = Field(min_length=1)
    exec_param: dict[str, Any] = Field(default_factory=dict)
    desc: Optional[str] = Field(default=None, max_length=1024)

File: fronted/src/api/apiAutomation.ts
import { del, get, post } from './http'
import type { AgentTask } from './agentTask'
import { getPageItems, type PaginatedData } from './pagination'

export interface ApiRequestDetails {
  path: string
  method: string
  url_params: Record<string, unknown>
  form: Record<string, unknown>
  json: Record<string, unknown>
  cookies: Record<string, unknown>
  headers: Record<string, unknown>
}

export interface ApiAutomationCase {
  id: number
  project_id?: number | null
  module_id?: number | null
  module_name: string
  exec_type: string
  priority: number
  name: string
  request: ApiRequestDetails
  expected: string
  created_at?: string | null
}

export interface ApiExecutionResult {
  case_id: number
  case_name: string
  status: string
  expected: string
  ai_record: string
  response: {
    status_code: number
    headers: Record<string, unknown>
    body: unknown
  }
  request: ApiRequestDetails
}

export interface ApiExecutionDetails {
  summary: {
    success: number
    failed: number
    total: number
  }
  results: ApiExecutionResult[]
}

export interface ApiAutomationExec {
  id: number
  project_id?: number | null
  name: string
  exec_type: string
  case_ids: number[]
  details?: ApiExecutionDetails | null
  desc?: string | null
  exec_param: Record<string, unknown>
  exec_status: string
  created_at?: string | null
  updated_at?: string | null
}

export interface ApiAutomationExecPayload {
  project_id?: number | null
  name: string
  exec_type: string
  case_ids: number[]
  exec_param: Record<string, unknown>
  desc?: string | null
}

export interface ApiAutomationCaseQuery {
  project_id?: number
  name?: string
  priority?: number
  module_id?: number
  exec_type?: string
}

export function getApiAutomationCases(
  params?: ApiAutomationCaseQuery,
): Promise<ApiAutomationCase[]> {
  return get<PaginatedData<ApiAutomationCase>>('/api/v1/api-automation/cases', params).then(
    getPageItems,
  )
}

export function getApiAutomationCase(caseId: number): Promise<ApiAutomationCase> {
  return get<ApiAutomationCase>(`/api/v1/api-automation/cases/${caseId}`)
}

export function getApiAutomationExecs(projectId?: number): Promise<ApiAutomationExec[]> {
  return get<PaginatedData<ApiAutomationExec>>(
    '/api/v1/api-automation/execs',
    projectId ? { project_id: projectId } : undefined,
  ).then(getPageItems)
}

export function createApiAutomationExec(
  payload: ApiAutomationExecPayload,
): Promise<AgentTask<ApiExecutionDetails>> {
  return post<AgentTask<ApiExecutionDetails>>('/api/v1/api-automation/execs', payload)
}

export function copyApiAutomationExec(execId: number): Promise<ApiAutomationExec> {
  return post<ApiAutomationExec>(`/api/v1/api-automation/execs/${execId}/copy`)
}

export function deleteApiAutomationExec(execId: number): Promise<ApiAutomationExec> {
  return del<ApiAutomationExec>(`/api/v1/api-automation/execs/${execId}`)
}
